aws/table_aws_lakeformation_permission.go

```diff
@@ -0,0 +1,359 @@
+		if limit < maxItems {
+			if limit < 1 {
+				maxItems = int32(1)
+			} else {
+				maxItems = int32(limit)
+			}
+		}
+	}
+	input.MaxResults = aws.Int32(maxItems)
+
+	resourceInput := buildLakeformationResourceInputFilter(ctx, d.Quals)
+	if resourceInput != nil {
+		input.Resource = resourceInput
+	}
+
+	// Error: aws: operation error LakeFormation: ListPermissions, https response error StatusCode: 400, RequestID: 4dd01db3-77a7-4fcc-86d2-398e7f62fbcf, InvalidInputException: Resource is mandatory if Principal is set in the input.
+	if d.EqualsQualString("principal_identifier") != "" && resourceInput != nil {
+		input.Principal = &types.DataLakePrincipal{
+			DataLakePrincipalIdentifier: aws.String(d.EqualsQualString("principal_identifier")),
+		}
+	}
+
+	paginator := lakeformation.NewListPermissionsPaginator(svc, input, func(o *lakeformation.ListPermissionsPaginatorOptions) {
+		o.Limit = maxItems
+		o.StopOnDuplicateToken = true
+	})
+
+	for paginator.HasMorePages() {
+		// Apply rate limiting
+		d.WaitForListRateLimit(ctx)
+
+		output, err := paginator.NextPage(ctx)
+		if err != nil {
+			plugin.Logger(ctx).Error("aws_lakeformation_permission.listLakeformationPermissions", "api_error", err)
+			return nil, err
+		}
+
+		for _, permission := range output.PrincipalResourcePermissions {
+			d.StreamListItem(ctx, permission)
+
+			// Stop if the context is cancelled
+			if d.RowsRemaining(ctx) == 0 {
+				return nil, nil
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func buildLakeformationResourceInputFilter(ctx context.Context, quals plugin.KeyColumnQualMap) (resource *types.Resource) {
+	resourceInput := &types.Resource{}
+	hasValues := false // Track if any value is set
+
+	for columnName := range quals {
+		if quals[columnName] != nil {
+			plugin.Logger(ctx).Error("Column Name ===>>", columnName)
+			value := getQualsValueByColumn(quals, columnName, "string")
+			val, ok := value.(string)
+			if !ok {
+				continue
+			}
+
+			switch columnName {
+			case "database_catalog_id":
+				// If we are providing database_catalog_id then we must have to provide database_name
+				// Otherwise we will get the error:
+				// Error: aws: operation error LakeFormation: ListPermissions, 1 validation error(s) found.
+				// - missing required field, ListPermissionsInput.Resource.Database.Name.
+				if quals["database_name"] != nil {
+					if resourceInput.Database == nil {
+						resourceInput.Database = &types.DatabaseResource{}
+					}
+					resourceInput.Database.CatalogId = aws.String(val)
+					hasValues = true
+				}
+			case "database_name":
+				if resourceInput.Database == nil {
+					resourceInput.Database = &types.DatabaseResource{}
+				}
+				resourceInput.Database.Name = aws.String(val)
+				hasValues = true
+
+			// Error: aws: operation error LakeFormation: ListPermissions, 1 validation error(s) found.
+			// - missing required field, ListPermissionsInput.Resource.DataLocation.ResourceArn.
+			//  (SQLSTATE HV000
```

Comment:
Do we need these comments?

Reply:
It would be helpful to include the actual error in the comments, as it would provide a clear understanding of why the following logic was added and the specific issue it addresses.